File: EstruturasFlexiveis/OutrasEstruturas/Stack/Java/Stack.java
package EstruturasFlexiveis.OutrasEstruturas.Stack.Java;

public class Stack<T> {
    //=====PRIVATE=====//
    private Node<T> first;
    private int length = 0;

    //=====CONSTRUCTOR=====//
    public Stack() {
        this.setFirst(null);
    }

    @SuppressWarnings("unchecked")
    public Stack(T... arr) {
        for(T obj : arr) 
            this.push(obj);
        invert();
    }

    //=====GET=====//
    private Node<T> getFirst() {
        return this.first;
    }

    public int size() {
        return this.length;
    }

    //=====SET=====//
    private void setFirst(Node<T> first) {
        this.first = first;
    }

    private void inc() {
        this.length++;
    }

    private void dec() {
        this.length--;
    }

    //=====METODOS=====//
    public Boolean isEmpty() {
        return this.size() == 0;
    }

    //=====PUSH=====//
    public void push(T obj) {
        Node<T> node  = new Node<>(obj, this.getFirst());
        this.setFirst(node);
        inc();
    }

    //=====POP=====//
    public T pop() {
        T content = this.getFirst().getObj();
        Node<T> curr = this.getFirst().getNext();
        this.getFirst().setNext(null);
        this.setFirst(curr);
        curr = null;
        dec();
        return content;
    }

    //=====PEEK=====//
    public T peek() {
        return this.getFirst().getObj();
    }

    //=====INVERT=====//
    @SuppressWarnings("unchecked")
    public void invert() {
        Object[] objs = new Object[this.size()];
        
        for(int i = 0; i < objs.length; i++)
            objs[i] = pop();

        for(int i = 0; i < objs.length; i++)
            push((T)objs[i]); 
    }

    //=====OVERRIDE=====//
    @Override
    public String toString() {
        StringBuffer sb = new StringBuffer("[");
        Node<T> curr = this.getFirst();
        while(curr != null) {
            sb.append(curr.getObj());
            if(curr.getNext() != null) sb.append(", ");
            curr = curr.getNext();
        }
        curr = null;
        return sb.append("]").toString();
    }
}

File: EstruturasFlexiveis/OutrasEstruturas/Stack/C/Stack.c

#include "header/Stack.h"

//=====CONSTRUCTOR=====//
Stack* construct() {
    Stack* stack = (Stack*)malloc(sizeof(Stack));
    clear(stack);
    return stack;
}

//=====INITIALIZER=====//
void clear(Stack* const stack) {
    stack->first = NULL;
    stack->last = NULL;
    stack->len = 0;
}

//=====DESTRUCTOR=====//
void destruct(Stack* const stack) {
    Node* node = stack->first;
    Node* prev = NULL;

    while(node != NULL) {
        prev = node;
        node = node->next;
        free(prev);
    }

    free(stack);
}

//=====METODOS=====//
//-----ISEMPTY-----//
bool isEmptys(const Stack* const stack) {
    return stack->len == 0;
}

//-----PUSH-----//
void push(Stack* const stack, int obj) {
    Node* node = constructNodeObj(obj);
    

    if(isEmptys(stack)) stack->first = stack->last = node;
    else {
        node->next = stack->first;
        stack->first = node;
    }
    stack->len++;
}

//-----POP-----//
int pop(Stack* const stack) {
    int obj = 0;

    if(!isEmptys(stack)) {
        obj = stack->first->obj;

        if(stack->first == stack->last) {
            free(stack->first);
            stack->first = stack->last = NULL;
        } else {
            Node* curr = stack->first;
            stack->first = stack->first->next;
            free(curr);
        }
        
        stack->len--;
    }

    return obj;
}

//-----PEEK-----//
int peek(const Stack* const stack) {
    return (!isEmptys(stack)) ? stack->first->obj : 0;
}

//-----PRINT-----//
void prints(const Stack* const stack) {
    printf("[");

    Node* node = stack->first;
    while(node != NULL) {
        printf("%i", node->obj);
        if(node->next != NULL) printf(", ");
        node = node->next;
    }

    printf("]\n");
}